src-self-hosted/main.zig

```diff
@@ -729,15 +729,37 @@ fn cmdFmt(allocator: &Allocator, args: []const []const u8) !void {
     };
 
     for (flags.positionals.toSliceConst()) |file_path| {
-        var file = try os.File.openRead(allocator, file_path);
-        defer file.close();
-
-        const source_code = io.readFileAlloc(allocator, file_path) catch |err| {
+        var file = os.File.openRead(allocator, file_path) catch |err| {
             try stderr.print("unable to open '{}': {}", file_path, err);
             continue;
         };
+        defer file.close();
+        const size = try file.getEndPos();
+
+        var adaptater = io.FileInStream.init(&file);
+
+        var source_code: []u8 = undefined;
+        var start_index: usize = 0;
+
+        if(size < 16){
+            source_code = try allocator.alloc(u8, size);
+        }
+        else {
+            source_code = try allocator.alloc(u8, 16);
+            start_index = try adaptater.stream.read(source_code);
+            if(mem.eql(u8, source_code, "// zig fmt: skip")){
+                try stderr.print("won't format {}\n", file_path);
+                allocator.free(source_code);
```

Comment:
Why do we need this free? Won't it happen at line 756? If you want to handle the potential `stderr` error, then use `errdefer allocator.free(source_code)`, or even better, move the `defer allocator.free(source_code)` to before this new code.

Reply:
I believe that `continue` on the next line will jump back to the top of the loop, without executing line 756.
I didn't move it to the beginning, because in the event that the allocation fails, I assume that `source_code` would still be `undefined`, and I didn't want to `free(undefined)`. 
I may be mistaken.

Reply:
Since it is a fixed size, one thing you could do here is read into a local scope `[16]u8` and if it doesn't match seek(0) on the file and then let the actual work happen. Seems like it would be cleaner since there's just one check and an early `continue`.

Reply:
Yes, I actually planned on doing that but couldn't find a seek function. Well, I checked again and found `seekTo` in `file.zig` and I now feel stupid :smiley: